### PR TITLE
feat: Bridge module and upgrade sdk50

### PR DESCRIPTION
**Pull request by trinitys7**

### CODE REVIEW

File: app/upgrades/sdk50/constants.go
Comment (neitdung) [inline app/upgrades/sdk50/constants.go:1]: should remove if we use v2 folder

**Comment by neitdung** [comment]
We should also upgrade the third-party proto @trinitys7.

File: app/app.go [hunk 1]
```diff
@@ -344,36 +382,88 @@ func New(
...
+		appCodec,
+		runtime.NewKVStoreService(keys[banktypes.StoreKey]),
+		app.AccountKeeper,
+		app.ModuleAccountAddrs(),
+		authtypes.NewModuleAddress(govtypes.ModuleName).String(),
+		logger,
 	)
-	stakingKeeper := stakingkeeper.NewKeeper(
-		appCodec, keys[stakingtypes.StoreKey], app.AccountKeeper, app.BankKeeper, app.GetSubspace(stakingtypes.ModuleName),
+	app.StakingKeeper = stakingkeeper.NewKeeper(
```
Comment (neitdung) [inline app/app.go:395]: StakingKeeper shouldn't be added to app, we have had MultiStakingKeeper

**Comment by facs95** [comment]
The linter is broken, please fix it 🙏

File: x/asset/types/keys.go [hunk 2]
```diff
@@ -36,12 +38,6 @@ func KeyPrefix(p string) []byte {
 // TokenKey returns the store key to retrieve a Token from the index fields
 func TokenKey(
```
Comment (lacsomot) [inline x/asset/types/keys.go:39]: Check again here.

File: x/asset/keeper/keeper.go [hunk 3]
```diff
@@ -1,58 +1,78 @@
...
+	"cosmossdk.io/log"
 
 	"github.com/cosmos/cosmos-sdk/codec"
 	sdk "github.com/cosmos/cosmos-sdk/types"
-	paramtypes "github.com/cosmos/cosmos-sdk/x/params/types"
 
 	"github.com/realiotech/realio-network/x/asset/types"
 )
 
 type (
```
Comment (neitdung) [inline x/asset/keeper/keeper.go]: should remove this `(`

**Comment by neitdung** [comment]
@trinitys7 upgrade module should be removed from `SetOrderBeginBlockers`: https://docs.cosmos.network/main/build/migrations/upgrading#baseapp-1

**Comment by catShaark** [comment]
> @trinitys7 upgrade module should be removed from `SetOrderBeginBlockers`: https://docs.cosmos.network/main/build/migrations/upgrading#baseapp-1

Just check the code, we've set preblocker logic as instructed, upgrade module beginblocker removed, everything seems good.

File: x/asset/keeper/msg_server_authorize_address.go [hunk 4]
```diff
@@ -10,23 +10,25 @@ import (
...
+		return nil, err
+	}
+
 	if len(signers) != 1 {
 		return nil, errorsmod.Wrap(sdkerrors.ErrUnauthorized, "invalid signers")
 	}
 
 	// assert that the manager account is the only signer of the message
-	if signers[0].String() != token.Manager {
+	if sdk.AccAddress(signers[0]).String() != token.Manager {
```
Comment (neitdung) [inline x/asset/keeper/msg_server_authorize_address.go]: Can we just compare `msg.Manager` with `token.Manager` and validate signer in function `GetSigner`

File: x/asset/keeper/msg_server_un_authorize_address.go [hunk 5]
```diff
@@ -9,23 +9,27 @@ import (
...
+		return nil, err
+	}
+
 	if len(signers) != 1 {
 		return nil, errorsmod.Wrap(sdkerrors.ErrUnauthorized, "invalid signers")
 	}
 
 	// assert that the manager account is the only signer of the message
-	if signers[0].String() != token.Manager {
+	if sdk.AccAddress(signers[0]).String() != token.Manager {
```
Comment (neitdung) [inline x/asset/keeper/msg_server_un_authorize_address.go]: Should replace this with msg.Manager

File: x/asset/keeper/msg_server_update_token.go [hunk 6]
```diff
@@ -10,22 +10,22 @@ import (
...
-	if len(signers) != 1 {
-		return nil, errorsmod.Wrap(sdkerrors.ErrUnauthorized, "invalid signers")
+	signers, _, err := ms.cdc.GetMsgV1Signers(msg)
+	if err != nil {
+		return nil, err
 	}
 
 	// assert that the manager account is the only signer of the message
-	if signers[0].String() != existing.Manager {
+	if sdk.AccAddress(signers[0]).String() != existing.Manager {
```
Comment (neitdung) [inline x/asset/keeper/msg_server_update_token.go]: Should replace with `msg.Manager`

**Comment by neitdung** [comment]
@trinitys7 @catShaark do we need to write a migrate function for converting `KVStore` to `collections`

File: x/bridge/types/params_legacy.go
Comment (neitdung) [inline x/bridge/types/params_legacy.go:1]: This file is not necessary. We just need this when we upgrade from sdk46 to sdk47

File: x/bridge/types/ratelimit.go [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+		return errorsmod.Wrap(ErrInflowThresholdExceeded,
+			fmt.Sprintf("tx amount (%v) exceeds threshold (%v)", amount.String(), r.Ratelimit.String()),
+		)
+	}
+
+	r.CurrentInflow = netInflow
+	return nil
+}
+
+func (r *RateLimit) ResetInflow() {
```
Comment (neitdung) [inline x/bridge/types/ratelimit.go]: This function just be called once, we should remove this

File: app/upgrades/v2/upgrades.go [hunk 8]
```diff
@@ -1,20 +1,177 @@
...
+		}
+		err = bridgeKeeper.RegisteredCoins.Set(ctx, "arst", bridgetypes.RateLimit{
+			Ratelimit:     math.Int(math.NewUintFromString("1000000000000000000000000")),
+			CurrentInflow: math.ZeroInt(),
+		})
+		if err != nil {
+			return nil, err
+		}
+		err = bridgeKeeper.EpochInfo.Set(ctx, bridgetypes.EpochInfo{
+			StartTime:            time.Unix(int64(1729763876), 0),
```
Comment (neitdung) [inline app/upgrades/v2/upgrades.go]: StartTime should not be fixed

Comment (trinitys7) [reply app/upgrades/v2/upgrades.go]: The start time is only set once in order to trigger the epoch counting. So it should be fixed base on the upgrade time.

File: x/asset/keeper/msg_server_authorize_address.go [hunk 9]
```diff
@@ -10,23 +10,25 @@ import (
...
-	if !isFound {
-		return nil, errorsmod.Wrapf(sdkerrors.ErrKeyNotFound, "symbol %s does not exists", msg.Symbol)
+	token, err := ms.Token.Get(goCtx, types.TokenKey(msg.Symbol))
+	if err != nil {
+		return nil, errorsmod.Wrapf(sdkerrors.ErrKeyNotFound, "symbol %s does not exists : %s", msg.Symbol, err.Error())
 	}
 
 	// Checks if the token manager signed
-	signers := msg.GetSigners()
+	signers, _, err := ms.cdc.GetMsgV1Signers(msg)
```
Comment (neitdung) [inline x/asset/keeper/msg_server_authorize_address.go]: Should remove this

File: x/asset/keeper/msg_server_un_authorize_address.go [hunk 10]
```diff
@@ -9,23 +9,27 @@ import (
...
-	if !isFound {
-		return nil, errorsmod.Wrapf(sdkerrors.ErrKeyNotFound, "symbol %s does not exists", msg.Symbol)
+	token, err := ms.Token.Get(ctx, types.TokenKey(msg.Symbol))
+	if err != nil {
+		return nil, errorsmod.Wrapf(sdkerrors.ErrKeyNotFound, "symbol %s does not exists: %s", msg.Symbol, err.Error())
 	}
 
 	// Checks if the token manager signed
-	signers := msg.GetSigners()
+	signers, _, err := ms.cdc.GetMsgV1Signers(msg)
```
Comment (neitdung) [inline x/asset/keeper/msg_server_un_authorize_address.go]: Should remove signer validation

**Comment by trinitys7** [comment]
> @trinitys7 @catShaark do we need to write a migrate function for converting `KVStore` to `collections`

No we don't, I tested them on mainnet state upgrade

File: app/ante/ante.go [hunk 11]
```diff
@@ -20,18 +22,18 @@ func NewAnteHandler(options HandlerOptions) sdk.AnteHandler {
 	) (newCtx sdk.Context, err error) {
 		var anteHandler sdk.AnteHandler
 
-		defer ethante.Recover(ctx.Logger(), &err)
+		defer Recover(ctx.Logger(), &err)
```
Comment (facs95) [inline app/ante/ante.go]: why is this recover needed?

File: app/ante/ante.go [hunk 12]
```diff
@@ -20,18 +22,18 @@ func NewAnteHandler(options HandlerOptions) sdk.AnteHandler {
...
 			opts := txWithExtensions.GetExtensionOptions()
 			if len(opts) > 0 {
 				switch typeURL := opts[0].GetTypeUrl(); typeURL {
-				case "/ethermint.evm.v1.ExtensionOptionsEthereumTx":
+				case "/os.evm.v1.ExtensionOptionsEthereumTx":
 					// handle as *evmtypes.MsgEthereumTx
 					anteHandler = newEthAnteHandler(options)
-				case "/ethermint.types.v1.ExtensionOptionsWeb3Tx":
-					// handle as normal Cosmos SDK tx, except signature is checked for EIP712 representation
+				case "/os.types.v1.ExtensionOptionDynamicFeeTx":
```
Comment (facs95) [inline app/ante/ante.go:29]: This is wrong. ExtensionOptionDynamicFeeTx is set to be handle by the `newCosmosAnteHandler` not by the legacyEIP712. we should not need the `newLegacyCosmosAnteHandlerEipe712` anymore.

File: app/app.go [hunk 13]
```diff
@@ -22,35 +36,57 @@ import (
...
-	evmtypes "github.com/evmos/ethermint/x/evm/types"
-	"github.com/evmos/ethermint/x/evm/vm/geth"
-	"github.com/evmos/ethermint/x/feemarket"
-	feemarketkeeper "github.com/evmos/ethermint/x/feemarket/keeper"
-	feemarkettypes "github.com/evmos/ethermint/x/feemarket/types"
+	evmosante "github.com/evmos/os/ante"
+	evmosanteevm "github.com/evmos/os/ante/evm"
+	osecp256k1 "github.com/evmos/os/crypto/ethsecp256k1"
+	srvflags "github.com/evmos/os/server/flags"
+	ethermint "github.com/evmos/os/types"
```
Comment (facs95) [inline app/app.go]: this probably should be renamed

File: app/app.go [hunk 14]
```diff
@@ -466,6 +560,9 @@ func New(
 	ibcRouter.AddRoute(ibctransfertypes.ModuleName, transferStack)
 	app.IBCKeeper.SetRouter(ibcRouter)
 
+	app.EvmKeeper.WithStaticPrecompiles(
```
Comment (facs95) [inline app/app.go]: I think the sdk50 upgarde is big enough already, we should not add precompiles yet, that should probably be on an independent updrage.

File: app/app.go [hunk 15]
```diff
@@ -861,22 +1077,83 @@ func (app *RealioNetwork) SimulationManager() *module.SimulationManager {
...
+func RealioSigVerificationGasConsumer(
+	meter storetypes.GasMeter, sig signing.SignatureV2, params authtypes.Params,
+) error {
+	pubkey := sig.PubKey
+	switch pubkey := pubkey.(type) {
+	case *osecp256k1.PubKey:
+		// Ethereum keys
+		meter.ConsumeGas(evmosante.Secp256k1VerifyCost, "ante verify: eth_secp256k1")
+		return nil
+	case *ethsecp256k1.PubKey:
```
Comment (facs95) [inline app/app.go:1102]: Why is this needed?

File: app/app.go [hunk 16]
```diff
@@ -861,22 +1077,83 @@ func (app *RealioNetwork) SimulationManager() *module.SimulationManager {
...
+		return errorsmod.Wrap(errortypes.ErrInvalidPubKey, "ED25519 public keys are unsupported")
+	case multisig.PubKey:
+		// Multisig keys
+		multisignature, ok := sig.Data.(*signing.MultiSignatureData)
+		if !ok {
+			return fmt.Errorf("expected %T, got, %T", &signing.MultiSignatureData{}, sig.Data)
+		}
+		return evmosante.ConsumeMultisignatureVerificationGas(meter, multisignature, pubkey, params, sig.Sequence)
+	default:
+		return authante.DefaultSigVerificationGasConsumer(meter, sig, params)
```
Comment (facs95) [inline app/app.go:1118]: why do we support cosmos's default?

File: app/encoding.go [hunk 17]
```diff
@@ -1,11 +1,70 @@
...
+	evmcryptocodec "github.com/evmos/os/crypto/codec"
+	"github.com/evmos/os/ethereum/eip712"
+	evmostypes "github.com/evmos/os/types"
+	erc20types "github.com/evmos/os/x/erc20/types"
+	evmtypes "github.com/evmos/os/x/evm/types"
+	ethcryptocodec "github.com/realiotech/realio-network/crypto/codec"
 )
 
 // MakeEncodingConfig creates the EncodingConfig for realio network
 func MakeEncodingConfig() params.EncodingConfig {
```
Comment (facs95) [inline app/encoding.go:26]: This should change after removing legacy eip712

**Review by facs95 — CHANGES_REQUESTED** [review]

**Comment by facs95** [comment]
PR should be block until fixing the ante handler issues. I stopped the review there giving that that change should have significant changes in multiple files. Once it is fixed I will continue the review.

File: app/app.go (hunk 16 above)
Comment (catShaark) [reply app/app.go:1118]: It's for supporting regular cosmos signing, we don't want to break existing accounts with cosmos key. Also compatiblity with relayer

File: app/app.go (hunk 15 above)
Comment (trinitys7) [reply app/app.go:1102]: For backward compability with acount that has ethermint key.

Comment (trinitys7) [reply app/app.go:1102]: The ethsecp256k1 Pubkey is the pubkey for ethermint one, while the osecp256k1 Pubkey is the pubkey for evmos os, as we move from ethermint to os, the ethermint key will still be store in local keyring.

File: app/encoding.go (hunk 17 above)
Comment (trinitys7) [reply app/encoding.go:26]: sorry, can you elaborate that ?

File: app/ante/ante.go (hunk 12 above)
Comment (trinitys7) [reply app/ante/ante.go:29]: Copied that!

File: app/encoding.go (hunk 17 above)
Comment (facs95) [reply app/encoding.go:26]: Sorry I made a mistake, confused the eip712 with the legacy eip712. This is no needed.

File: app/app.go (hunk 16 above)
Comment (facs95) [reply app/app.go:1118]: I see, Evmos does not support that, I wonder if there are any security issues we havent thought about. (I dont have anything in mind though)

File: app/app.go (hunk 15 above)
Comment (facs95) [reply app/app.go:1102]: you are right, I can ask the evmos team if they can support the ethermint type so we dont have to support mulutiple types

File: app/upgrades/v2/upgrades.go [hunk 18]
```diff
@@ -1,20 +1,191 @@
...
+			Duration:             time.Minute,
+			EpochCountingStarted: false,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		evmParams := evmtypes.DefaultParams()
+		evmParams.ActiveStaticPrecompiles = []string{
```
Comment (facs95) [inline app/upgrades/v2/upgrades.go]: I think this is not needed is it?

File: app/upgrades/v2/upgrades.go [hunk 19]
```diff
@@ -1,20 +1,191 @@
...
+		err = bridgeKeeper.EpochInfo.Set(ctx, bridgetypes.EpochInfo{
+			StartTime:            time.Unix(int64(1729763876), 0),
+			Duration:             time.Minute,
+			EpochCountingStarted: false,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		evmParams := evmtypes.DefaultParams()
```
Comment (facs95) [inline app/upgrades/v2/upgrades.go:130]: Also, we should set the params so that evm module is permissioned

File: app/upgrades/v2/upgrades.go (hunk 18 above)
Comment (trinitys7) [reply app/upgrades/v2/upgrades.go]: Thank you, I will remove it.

**Review by facs95 — APPROVED** [review]

**Review by jiujiteiro — APPROVED** [review]
lgtm
